### PR TITLE
feat(VCalendar): Events improvements

### LINKED ISSUE #6264: [Feature Request] VCalendar: Overlapping events
### Problem to solve
It seems like V-Calendar's event template only takes into account the fact that events cannot overlap: the example sets a full width for every event, which prevents overlapping events from display properly (i.e. one on top of the other, instead of one next to each other).

### Proposed solution
Is there any plan to integrate the width/x calculation inside V-Calendar, or is it something we're supposed to implement ourselves?

### LINKED ISSUE #6307: [Feature Request] Calendar to support multi-day events
### Problem to solve
This will allow for the UI to show events in the calendar that span more than one day.

### Proposed solution
Follow how Google have implemented it in their calendar:

image

### PR DESCRIPTION
**Pull request by ClickerMonkey**
## Description
- Add events\* props to the calendar
- Visual changes to calendar to more closely match Google's
- Unit Tests

## Motivation and Context
fixes #6264 
fixes #6307 

- Events are highly requested
- The visual changes vastly improve the look of the calendar
- Unit tests were needed

## How Has This Been Tested?

unit, visually

## Markup:

<details>

```vue
<template>
  <v-sheet class="v-fullscreen" color="grey lighten-1">
    <v-toolbar fixed flat app color="white">
      <v-btn outline @click="setToday" class="mr-3">
        Today
      </v-btn>
      <v-btn fab text small @click="prev">
        <v-icon small>arrow_back_ios</v-icon>
      </v-btn>
      <v-btn fab text small @click="next">
        <v-icon small>arrow_forward_ios</v-icon>
      </v-btn>
      <v-toolbar-title>{{ title }}</v-toolbar-title>
      <v-spacer></v-spacer>
      <v-menu bottom right>
        <v-btn outline slot="activator">
          <span>{{ typeToLabel[type] }}</span>
          <v-icon right>arrow_drop_down</v-icon>
        </v-btn>
        <v-list>
          <v-list-tile @click="type = 'day'">
            <v-list-tile-title>Day</v-list-tile-title>
          </v-list-tile>
          <v-list-tile @click="type = 'week'">
            <v-list-tile-title>Week</v-list-tile-title>
          </v-list-tile>
          <v-list-tile @click="type = 'month'">
            <v-list-tile-title>Month</v-list-tile-title>
          </v-list-tile>
          <v-list-tile @click="type = '4day'">
            <v-list-tile-title>4 days</v-list-tile-title>
          </v-list-tile>
        </v-list>
      </v-menu>
    </v-toolbar>
    <v-calendar
      ref="calendar"
      color="primary"
      :events="events"
      :event-color="getEventColor"
      :now="today"
      :type="type"
      v-model="focus"
      @click:event="showEvent"
      @click:more="viewDay"
      @click:date="viewDay"
      @change="updateRange"
    ></v-calendar>
    <v-menu
      v-model="selectedOpen"
      :close-on-content-click="false"
      :activator="selectedElement"
      full-width
      offset-x
    >
      <v-card
        color="grey lighten-4"
        min-width="350px"
        flat
      >
        <v-toolbar
          :color="selectedEvent.color"
          dark
        >
          <v-btn icon>
            <v-icon>edit</v-icon>
          </v-btn>
          <v-toolbar-title v-html="selectedEvent.name"></v-toolbar-title>
          <v-spacer></v-spacer>
          <v-btn icon>
            <v-icon>favorite</v-icon>
          </v-btn>
          <v-btn icon>
            <v-icon>more_vert</v-icon>
          </v-btn>
        </v-toolbar>
        <v-card-title primary-title>
          <span v-html="selectedEvent.details"></span>
        </v-card-title>
        <v-card-actions>
          <v-btn
            text
            color="secondary"
          >
            Cancel
          </v-btn>
        </v-card-actions>
      </v-card>
    </v-menu>
  </v-sheet>
</template>

<script>
export default {
  data: () => ({
    today: '2019-01-08',
    focus: '2019-01-08',
    type: 'month',
    typeToLabel: {
      'month': 'Month',
      'week': 'Week',
      'day': 'Day',
      '4day': '4 Days'
    },
    start: null,
    end: null,
    selectedEvent: {},
    selectedElement: null,
    selectedOpen: false,
    events: [
      {
        name: 'Vacation',
        details: 'Going to the beach!',
        start: '2018-12-29',
        end: '2019-01-01',
        color: 'blue'
      },
      {
        name: 'Meeting',
        details: 'Spending time on how we do not have enough time',
        start: '2019-01-07 09:00',
        end: '2019-01-07 09:30',
        color: 'indigo'
      },
      {
        name: 'Large Event',
        details: 'This starts in the middle of an event and spans over multiple events',
        start: '2018-12-31',
        end: '2019-01-04',
        color: 'deep-purple'
      },
      {
        name: '3rd to 7th',
        details: 'Testing',
        start: '2019-01-03',
        end: '2019-01-07',
        color: 'cyan'
      },
      {
        name: 'Big Meeting',
        details: 'A very important meeting about nothing',
        start: '2019-01-07 08:00',
        end: '2019-01-07 11:30',
        color: 'red'
      },
      {
        name: 'Another Meeting',
        details: 'Another important meeting about nothing',
        start: '2019-01-07 10:00',
        end: '2019-01-07 13:30',
        color: 'brown'
      },
      {
        name: '7th to 8th',
        start: '2019-01-07',
        end: '2019-01-08',
        color: 'blue'
      },
      {
        name: 'Lunch',
        details: 'Time to feed',
        start: '2019-01-07 12:00',
        end: '2019-01-07 15:00',
        color: 'deep-orange'
      },
      {
        name: '30th Birthday',
        details: 'Celebrate responsibly',
        start: '2019-01-03',
        color: 'teal'
      },
      {
        name: 'New Year',
        details: 'Eat chocolate until you pass out',
        start: '2019-01-01',
        end: '2019-01-02',
        color: 'green'
      },
      {
        name: 'Conference',
        details: 'The best time of my life',
        start: '2019-01-21',
        end: '2019-01-28',
        color: 'grey darken-1'
      },
      {
        name: 'Hackathon',
        details: 'Code like there is no tommorrow',
        start: '2019-01-30 23:00',
        end: '2019-02-01 08:00',
        color: 'black'
      },
      {
        name: 'event 1',
        start: '2019-01-14 18:00',
        end: '2019-01-14 19:00',
        color: '#4285F4'
      },
      {
        name: 'event 2',
        start: '2019-01-14 18:00',
        end: '2019-01-14 19:00',
        color: '#4285F4'
      },
      {
        name: 'event 5',
        start: '2019-01-14 18:00',
        end: '2019-01-14 19:00',
        color: '#4285F4'
      },
      {
        name: 'event 3',
        start: '2019-01-14 18:30',
        end: '2019-01-14 20:30',
        color: '#4285F4'
      },
      {
        name: 'event 4',
        start: '2019-01-14 19:00',
        end: '2019-01-14 20:00',
        color: '#4285F4'
      },
      {
        name: 'event 6',
        start: '2019-01-14 21:00',
        end: '2019-01-14 23:00',
        color: '#4285F4'
      },
      {
        name: 'event 7',
        start: '2019-01-14 22:00',
        end: '2019-01-14 23:00',
        color: '#4285F4'
      }
    ]
  }),
  computed: {
    title () {
      const { start, end } = this
      if (!start || !end) {
        return ''
      }

      const startMonth = this.monthFormatter(start)
      const endMonth = this.monthFormatter(end)
      const suffixMonth = startMonth === endMonth ? '' : endMonth

      const startYear = start.year
      const endYear = end.year
      const suffixYear = startYear === endYear ? '' : endYear

      const startDay = start.day + this.nth(start.day)
      const endDay = end.day + this.nth(end.day)

      switch (this.type) {
        case 'month':
          return `${startMonth} ${startYear}`
        case 'week':
        case '4day':
          return `${startMonth} ${startDay} ${startYear} - ${suffixMonth} ${endDay} ${suffixYear}`
        case 'day':
          return `${startMonth} ${startDay} ${startYear}`
      }
      return ''
    },
    monthFormatter () {
      return this.$refs.calendar.getFormatter({
        timeZone: 'UTC', month: 'long'
      })
    }
  },
  methods: {
    viewDay ({date}) {
      this.focus = date
      this.type = 'day'
    },
    getEventColor (event) {
      return event.color
    },
    setToday () {
      this.focus = this.today
    },
    prev () {
      this.$refs.calendar.prev()
    },
    next () {
      this.$refs.calendar.next()
    },
    showEvent ({ nativeEvent, event }) {
      const open = () => {
        this.selectedEvent = event
        this.selectedElement = nativeEvent.target
        this.selectedOpen = true
      }

      if (this.selectedOpen) {
        this.selectedOpen = false 
        setTimeout(open, 10)
      } else {
        open()
      }
      
      nativeEvent.stopPropagation();
    },
    updateRange ({ start, end }) {
      // You could load events from an outside source (like database) now that we have the start and end dates on the calendar
      this.start = start
      this.end = end
    },
    nth(d) {
      return d > 3 && d < 21
        ? 'th'
        : ['th', 'st', 'nd', 'rd', 'th', 'th', 'th', 'th', 'th', 'th'][d % 10]
    }
  }
}
</script>

<style scoped>
  .v-fullscreen {
    position: absolute;
    padding-top: 1px;
    top: 0px;
    left: 0px;
    right: 0px;
    bottom: 0px;
  }
</style>
```
</details>

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any feature but make things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and breaking changes).
- [x] My code follows the code style of this project.
- [x] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

### CODE REVIEW

**Comment by ClickerMonkey** [comment]
I've created the PR awhile to keep track of changes and related issues.

**Comment by KaelWD** [comment]
The test coverage seems fine if you actually look at the diff, it's just the PR bot is showing nonsense for some reason.

**Comment by rkacenski** [comment]
Hey, really interested in seeing this merged in. I went ahead made some changes on a fork:
https://github.com/rkacenski/vuetify/commit/72c8fa755c58237f249dd3b46a9e172b7b890594

The changes made are:
- fixed the sass bugs in monthly view
- matched the day interger on the month view to the google calendar style
- multi-day events are spanning over days without padding
- fixed a few bugs in the docs example
- added  a new prop to fix a bug with the margin and positioning of multi-day events, 'eventMarginBottom'

Not sure whats the best way to get these changes into this branch, please advise. Thanks

**Comment by ClickerMonkey** [comment]
@rkacenski (most if not all) of those changes were already present in the code - so hopefully your code didn't overwrite anything important or more closely styled to Google Calendar.

**Comment by ClickerMonkey** [comment]
I looked over your code now, makes sense what you did. I thought you meant you implemented it your own way and try to merge it into the code base which already had many of those changes.

**Comment by MajesticPotatoe** [comment]
Fixing conflicts

**Comment by rkacenski** [comment]
@ClickerMonkey oh nah, I was just trying to fix up the stuff that was broken. You did some good work on this, just wanted to help get in merged in.

**Comment by rkacenski** [comment]
Also I fixed the test, they run fine now. Just having one small issue with: Type 'null' is not assignable to type 'VTimestamp'. Could use some help solving it, more info on the PR:
https://github.com/vuetifyjs/vuetify/pull/7408

**Comment by rkacenski** [comment]
@MajesticPotatoe should be good to go now

File: packages/docs/src/examples/calendars/playground.vue [hunk 1]
```diff
@@ -73,34 +73,34 @@
       >
         <template v-slot:activator="{ on }">
           <v-text-field
+            slot="activator"
```
Comment (DRoet) [inline packages/docs/src/examples/calendars/playground.vue]: conflict: should use the new v-slot behaviour

File: packages/docs/src/examples/calendars/playground.vue [hunk 2]
```diff
@@ -111,33 +111,32 @@
...
-          <v-text-field
-            v-model="end"
-            label="End Date"
-            prepend-icon="event"
-            readonly
-            v-on="on"
-          ></v-text-field>
-        </template>
+        <v-text-field
+          slot="activator"
```
Comment (DRoet) [inline packages/docs/src/examples/calendars/playground.vue]: conflict: should use the new v-slot behaviour

File: packages/docs/src/examples/calendars/playground.vue [hunk 3]
```diff
@@ -153,33 +152,32 @@
...
-          <v-text-field
-            v-model="now"
-            label="Today"
-            prepend-icon="event"
-            readonly
-            v-on="on"
-          ></v-text-field>
-        </template>
+        <v-text-field
+          slot="activator"
```
Comment (DRoet) [inline packages/docs/src/examples/calendars/playground.vue]: conflict: should use the new v-slot behaviour

File: packages/docs/src/examples/expansion-panels/simple/disabled.vue [hunk 4]
```diff
@@ -41,6 +41,7 @@
     data: () => ({
       panel: [0, 1],
       disabled: false,
+      readonly: false,
```
Comment (DRoet) [inline packages/docs/src/examples/expansion-panels/simple/disabled.vue]: seems unrelated to the PR, no?

**Comment by DRoet** [comment]
@johnleider could this still make its way into 2.0? kinda related to MD2 and this PR has been open for a while

File: packages/vuetify/src/components/VCalendar/VCalendarWeekly.sass [hunk 5]
```diff
@@ -52,27 +55,23 @@
...
   box-shadow: none
   text-align: center
-  left: 0
-  top: 0
-  border-radius: $calendar-weekly-day-label-radius
-  width: $calendar-weekly-day-label-size
-  height: $calendar-weekly-day-label-size
-  line-height: $calendar-weekly-day-label-size
+  margin: 4px 0 0 0
+  //border-radius: $calendar-weekly-day-label-radius
```
Comment (sh7dm) [inline packages/vuetify/src/components/VCalendar/VCalendarWeekly.sass]: remove comments?

File: packages/vuetify/src/components/VCalendar/VCalendarWeekly.sass [hunk 6]
```diff
@@ -52,27 +55,23 @@
...
   cursor: pointer
   box-shadow: none
   text-align: center
-  left: 0
-  top: 0
-  border-radius: $calendar-weekly-day-label-radius
-  width: $calendar-weekly-day-label-size
-  height: $calendar-weekly-day-label-size
-  line-height: $calendar-weekly-day-label-size
+  margin: 4px 0 0 0
```
Comment (sh7dm) [inline packages/vuetify/src/components/VCalendar/VCalendarWeekly.sass]: move to variable?

File: packages/vuetify/src/components/VCalendar/_variables.scss [hunk 7]
```diff
@@ -1,15 +1,19 @@
...
-$calendar-weekly-weekday-font-size: 14px !default;
-$calendar-weekly-day-padding: 32px 4px 4px 4px !default;
+$calendar-weekly-weekday-font-size: 11px !default;
+$calendar-weekly-day-padding: 0px 0px 0px 0px !default;
 $calendar-weekly-day-label-size: 32px !default;
 $calendar-weekly-day-label-radius: 16px !default;
+$calendar-weekly-day-label-font-size: 12px !default;
 $calendar-weekly-day-month-left: 36px !default;
+
+$calendar-event-bottom-space: 1px;
```
Comment (sh7dm) [inline packages/vuetify/src/components/VCalendar/_variables.scss]:
```suggestion
$calendar-event-bottom-space: 1px !default;
```

File: packages/vuetify/src/components/VCalendar/_variables.scss [hunk 8]
```diff
@@ -1,15 +1,19 @@
...
+$calendar-event-border-radius: 4px;
```
Comment (sh7dm) [inline packages/vuetify/src/components/VCalendar/_variables.scss]:
```suggestion
$calendar-event-border-radius: 4px !default;
```

File: packages/vuetify/src/components/VCalendar/_variables.scss [hunk 9]
```diff
@@ -1,15 +1,19 @@
...
+$calendar-event-font-size: 12px;
```
Comment (sh7dm) [inline packages/vuetify/src/components/VCalendar/_variables.scss]:
```suggestion
$calendar-event-font-size: 12px !default;
```

File: packages/vuetify/src/components/VCalendar/mixins/__tests__/calendar-with-events.spec.ts [hunk 10]
```diff
@@ -0,0 +1,504 @@
...
+import CalendarWithEvents from '../calendar-with-events'
+import { parseTimestamp } from '../../util/timestamp'
+import { parseEvent } from '../../util/events'
+
+const Mock = CalendarWithEvents.extend({
+  render: h => h('div'),
+})
+
+describe('calendar-with-events.ts', () => {
+  type Instance = ExtractVue<typeof Mock>
```
Comment (sh7dm) [inline packages/vuetify/src/components/VCalendar/mixins/__tests__/calendar-with-events.spec.ts]:
```suggestion
  type Instance = InstanceType<typeof Mock>
```

File: packages/vuetify/src/components/VCalendar/mixins/__tests__/calendar-with-events.spec.ts [hunk 11]
```diff
@@ -0,0 +1,504 @@
+import {
+  mount,
+  Wrapper,
+  MountOptions,
+} from '@vue/test-utils'
+import { ExtractVue } from '../../../../util/mixins'
```
Comment (sh7dm) [inline packages/vuetify/src/components/VCalendar/mixins/__tests__/calendar-with-events.spec.ts]:
```suggestion
```

File: packages/vuetify/src/components/VCalendar/mixins/calendar-base.ts [hunk 12]
```diff
@@ -30,6 +33,10 @@ export default mixins(
 ).extend({
   name: 'calendar-base',
 
+  directives: {
+    Resize,
```
Comment (sh7dm) [inline packages/vuetify/src/components/VCalendar/mixins/calendar-base.ts]: is it used anywhere here?

File: packages/vuetify/src/components/VCalendar/mixins/calendar-with-events.sass [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+.v-calendar
+  .v-event
+    position: relative
+    overflow: hidden
+    text-overflow: ellipsis
+    white-space: nowrap
+    font-size: $calendar-event-font-size
+    cursor: pointer
+    margin-right: -1px
+    &.v-event-start
```
Comment (sh7dm) [inline packages/vuetify/src/components/VCalendar/mixins/calendar-with-events.sass]: Please format this and :point_down: with blank lines for clarity like in other styles.

File: packages/vuetify/src/components/VCalendar/mixins/calendar-with-events.ts [hunk 14]
```diff
@@ -0,0 +1,475 @@
...
+      const hour = withTime.hour % 12 || 12
+      const minute = withTime.minute
+
+      return minute > 0
+        ? (minute < 10
+          ? `${hour}:0${minute}${suffix}`
+          : `${hour}:${minute}${suffix}`)
+        : `${hour}${suffix}`
+    },
+    updateEventVisibility (): void {
```
Comment (sh7dm) [inline packages/vuetify/src/components/VCalendar/mixins/calendar-with-events.ts]: IMHO no need for void types
```suggestion
    updateEventVisibility () {
```

File: packages/vuetify/src/components/VCalendar/mixins/calendar-with-events.ts [hunk 15]
```diff
@@ -0,0 +1,475 @@
...
+
+        if (hide) {
+          more.style.display = ''
+          more.innerHTML = this.$vuetify.lang.t(this.eventMoreText, hidden)
+        } else {
+          more.style.display = 'none'
+        }
+      }
+    },
+    hideEvents (id: string): void {
```
Comment (sh7dm) [inline packages/vuetify/src/components/VCalendar/mixins/calendar-with-events.ts]:
```suggestion
    hideEvents (id: string) {
```

File: packages/vuetify/src/components/VCalendar/util/events.ts [hunk 16]
```diff
@@ -0,0 +1,47 @@
+
```
Comment (sh7dm) [inline packages/vuetify/src/components/VCalendar/util/events.ts]:
```suggestion
```

File: packages/vuetify/src/components/VCalendar/util/events.ts [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+import {
+  VTimestamp,
+  parseTimestamp,
+  getDayIdentifier,
+  getTimestampIdentifier,
+} from './timestamp'
+
+export interface VEventInput {
+  [prop: string]: any
```
Comment (sh7dm) [inline packages/vuetify/src/components/VCalendar/util/events.ts]: maybe we need a more strong type here

File: packages/vuetify/src/components/VCalendar/util/props.ts [hunk 18]
```diff
@@ -104,6 +105,56 @@ export default {
       validate: validateTimestamp,
     },
   },
+  events: {
+    events: {
+      type: Array as () => VEventInput[],
```
Comment (sh7dm) [inline packages/vuetify/src/components/VCalendar/util/props.ts]: `as (() => VEventInput)[],`??? :man_shrugging:
